hail/src/main/scala/is/hail/expr/types/encoded/ENDArray.scala

```diff
@@ -0,0 +1,112 @@
+package is.hail.expr.types.encoded
+
+import is.hail.annotations.{Region, StagedRegionValueBuilder}
+import is.hail.asm4s._
+import is.hail.expr.types.physical._
+import is.hail.expr.types.virtual._
+import is.hail.io.{InputBuffer, OutputBuffer}
+import is.hail.utils._
+
+final case class ENDArray(elementType: EType, override val required: Boolean = false) extends EType {
+  type DecodedPType = PCanonicalNDArray
+
+  override def decodeCompatible(pt: PType): Boolean = {
+    pt.required == required &&
+      pt.isInstanceOf[DecodedPType] &&
+      pt.asInstanceOf[DecodedPType].nDims == 2 &&
+      elementType.decodeCompatible(pt.asInstanceOf[DecodedPType].elementType)
+  }
+
+  override def encodeCompatible(pt: PType): Boolean = {
+    pt.required == required &&
+      pt.isInstanceOf[DecodedPType] &&
+      pt.asInstanceOf[DecodedPType].nDims == 2 &&
```

Comment:
I understand that this is intended to work specifically with block matrix partitions. Given that, should this class really be named `ENDArray`? I haven't look at other ETypes much, but this seems like a particular specialized NDArray etype, not the "canonical" one. Maybe like `ENDArrayMatrixBlock` or something?